File: frontend/src/types/graph.ts
// 图谱类型定义

export interface GraphNode {
  id: string;
  label: string;
  type: "person" | "entity" | "event";
  confidence: number;
  summary?: string;
  image_url?: string;
  year?: number;
}

export interface GraphEdge {
  source: string;
  target: string;
  type: string;
  confidence: number;
  source_url: string;
  evidence?: string;
}

export interface GraphResponse {
  center: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
  depth: number;
  has_more: boolean;
}

export type NodeType = "person" | "entity" | "event";
export type ConfidenceLevel = "high" | "medium" | "low";

// D3 力导向图扩展类型
export interface SimulationNode extends GraphNode {
  x?: number;
  y?: number;
  vx?: number;
  vy?: number;
  fx?: number | null;
  fy?: number | null;
}

export interface SimulationLink {
  source: string | SimulationNode;
  target: string | SimulationNode;
  type: string;
  confidence: number;
  source_url: string;
  evidence?: string;
}

File: backend/app/models/graph.py
from typing import Optional

from pydantic import BaseModel, Field


class GraphNode(BaseModel):
    id: str
    label: str
    type: str = "entity"  # person/entity/event/concept/technology/organization/category
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    relevance: float = Field(default=0.0, ge=0.0, le=1.0, description="与中心实体的相关度（≠confidence 数据可靠度）")
    summary: Optional[str] = None
    image_url: Optional[str] = None
    year: Optional[int] = None
    year_end: Optional[int] = None
    hop: Optional[int] = None  # 距中心跳数（多跳构建用；depth=1 时缺省）


class GraphEdge(BaseModel):
    source: str
    target: str
    type: str
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    relevance: float = Field(default=0.0, ge=0.0, le=1.0, description="与中心实体的相关度（≠confidence 数据可靠度）")
    source_url: str = ""
    evidence: Optional[str] = None
    hop: Optional[int] = None  # 这条边所属跳数（多跳构建用）


class GraphResponse(BaseModel):
    center: str
    nodes: list[GraphNode] = []
    edges: list[GraphEdge] = []
    depth: int = 1
    has_more: bool = False
